File: wordle/wordle.jl
using ArgParse
using ProgressMeter
using Combinatorics
using Random

function parse_arguments(args=ARGS)
    s = ArgParseSettings()
    @add_arg_table! s begin
        "solutions"
            help = "file containing solutions to Wordle"
            required = true
        "guessables"
            help = "file containing guessable words"
            required = true
        "--guesses","-g"
            help = "(exact) number of guesses to optimize"
            arg_type = Int
            default = 2
    end

    parsed_args = parse_args(args, s)

    return parsed_args
end

struct Word
    letters::NTuple{5, UInt8}
end

function Word(word::String)
    @inbounds Word((word[1], word[2], word[3], word[4], word[5]) .- '`')
end

letters(w::Word) = w.letters

function Base.show(io::IO, x::Word)
    show(io, String(collect(x.letters .+ '`')))
end

function Base.print(io::IO, x::Word)
    print(io, String(collect(x.letters .+ '`')))
end

import Base.getindex
function Base.getindex(w::Word, i)
    return getindex(w.letters, i)
end

function Base.getindex(w::Word, cs::NTuple{5, Bool})
    return getindex(w.letters, collect(cs))
end

function Base.isdisjoint(x::Word, y::Word)
    return isdisjoint(x.letters, y.letters)
end

function Base.isdisjoint(x::Vector{Word})
    isempty(x) && return true
    return length(unique(mapreduce(y->collect(letters(y)), vcat, x))) == length(x)
end

struct Wordle
    posmap::NTuple{5,NTuple{26,BitSet}}
    lettermap::NTuple{26,BitSet}
    missingmap::NTuple{26,BitSet}
    words::Vector{Word}
end

function nwords(w::Wordle)
    return length(w.words)
end

function getword(w::Wordle, i)
    return w.words[i]
end

function wordswithletter(w::Wordle, c::UInt8)
    return @inbounds w.lettermap[c]
end

function wordsmissingletter(w::Wordle, c::UInt8)
    return @inbounds w.missingmap[c]
end

function wordsmatching(w::Wordle, c::Pair{Int,UInt8})
    return @inbounds w.posmap[c[1]][c[2]]
end

const ALPHABET = "abcdefghijklmnopqrstuvwxyz"

function Wordle(words)
    lettermap = ntuple(x->BitSet(), 26)
    missingmap = ntuple(x->BitSet(), 26)
    posmap = ntuple(y->ntuple(x -> BitSet(), 26), 5)
    for i in eachindex(words)
        for c in ALPHABET
            char = c - 'a' + 1
            if c in words[i]
                push!(lettermap[char], i)
                for pos in findall(c, words[i])
                    push!(posmap[pos][char], i)
                end
            else
                push!(missingmap[char], i)
            end
        end
    end

    return Wordle(posmap, lettermap, missingmap, Word.(words))
end

function Base.show(io::IO, w::Wordle)
    show(io, w.words)
end

function guess(target::Word, g::Word)
    matches = g.letters .== target.letters
    contains = in.(g.letters, (target.letters,))
    return [i=>g[i] for i in findall(matches)], [g[i] for i in findall(contains)], [g[i] for i in findall(.!contains)]
end

function emptyintersect(s, itrs...)
    isempty(s) && return intersect(itrs...)
    return intersect(s, itrs...)
end

function possibilities(w::Wordle, matches::Vector{Pair{Int,UInt8}}, contains::Vector{UInt8}, misses::Vector{UInt8})
    ma = mapreduce((c)->wordsmatching(w, c), emptyintersect, matches; init=BitSet())
    co = mapreduce((c)->wordswithletter(w, c), emptyintersect, contains; init=ma)
    return mapreduce((c)->wordsmissingletter(w, c), emptyintersect, misses; init=co)
end

function ambiguities(w::Wordle, target::Word, gs::Vector{Word})
    possible = mapreduce(x->possibilities(w, guess(target, x)...), intersect, gs)
    1 / length(possible)
end

struct ComboProb
    combo::Vector{Word}
    prob::Float64
end

combination(c::ComboProb) = c.combo
probability(c::ComboProb) = c.prob

const BEST_COMBOPROB_SO_FAR = ComboProb(Word[], 0.)
const WORDS = Channel{Vector{Word}}(100)
const UNFILTERED_COMBOPROBS = Channel{ComboProb}(100)
const FILTERED_COMBOPROBS = Channel{ComboProb}(Inf)

function comboprob_filter()
    for combo in UNFILTERED_COMBOPROBS
        if probability(combo) > probability(best)
            BEST_COMBOPROB_SO_FAR = combo
            put!(FILTERED_COMBOPROBS, combo)
        end
    end
    close(FILTERED_COMBOPROBS)
end

function calculate_comboprob(wordle::Wordle, solutions::Vector{Word})
    for combo in WORDS
        if !isdisjoint(combo)
            continue
        end
        prob = mapreduce(x -> ambiguities(wordle, x, combo), +, solutions)
        comboprob = ComboProb(combo, prob)
        put!(UNFILTERED_COMBOPROBS, comboprob)
    end
    close(UNFILTERED_COMBOPROBS)
end

function main(args=ARGS)
    a = parse_arguments(args)
    @info "starting" args=args parsed_args=a

    solutions = readlines(a["solutions"])
    guessables = Word.(readlines(a["guessables"]))

    wordle = Wordle(solutions)
    combs = combinations(guessables, a["guesses"])
    ncombs = length(combs)
    # ncombs = 100
    p = Progress(ncombs; showspeed=true)
    solutions = Word.(solutions)

    for i in 1:Threads.nthreads()
        @info "We starting this?" i
        @async errormonitor(calculate_comboprob($wordle, $solutions))
    end
    @info "Is this starting?"
    @async errormonitor(comboprob_filter())

    for combo in combs
        println("trying $combo")
        put!(WORDS, combo)
        next!(p; showvalues=[(:best_combo,join(combination(BEST_COMBOPROB_SO_FAR), " + ")), (:prob, probability(BEST_COMBOPROB_SO_FAR)/length(solutions))])
    end

    close(WORDS)
    # deal with the history of the process to make sure we have the correct values
    for result in FILTERED_COMBOPROBS
        println("$(join(combination(result), " + ")) => $(probability(result))")
    end

    return
end

# combs, ambicount = main();
main()

File: wordle/wordle.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"io"
	"log"
	"os"
	"strings"
	"sync"

	"github.com/cheggaaa/pb/v3"
	"github.com/kelindar/bitmap"
	"gonum.org/v1/gonum/stat/combin"
)

var nGuesses = flag.Int("g", 2, "(exact) number of guesses to optimize after starting guesses")
var forceDisjoint = flag.Bool("d", false, "force all words in guess to have unique letters, including forbidding guesses to have duplicate letters")
var startingWords = flag.String("s", "", "comma-separated list of starting guesses")

const alphabetSize = 26
const wordSize = 5
const firstLetter = 'a'

type Word [wordSize]byte

func makeWord(bs []byte) Word {
	var w Word
	for i, b := range bs {
		if i >= wordSize {
			break
		}
		w[i] = b - firstLetter
	}
	return w
}

func makeWordFromString(s string) Word {
	return makeWord([]byte(s))
}

func (w Word) String() string {
	b := make([]byte, wordSize)
	for i := 0; i < wordSize; i++ {
		b[i] = w[i] + firstLetter
	}
	return string(b)
}

type Wordle struct {
	words []Word

	wordsContainingLetter           [alphabetSize]bitmap.Bitmap
	wordsContainingLetterByPosition [wordSize][alphabetSize]bitmap.Bitmap
	wordsNotContainingLetter        [alphabetSize]bitmap.Bitmap
}

func NewWordle(words []Word) *Wordle {
	ws := make([]Word, len(words))
	copy(ws, words)

	wordle := Wordle{words: ws}
	for i, word := range words {
		for letter := byte(0); letter < alphabetSize; letter++ {
			found := false
			for pos, character := range word {
				if character == letter {
					wordle.wordsContainingLetterByPosition[pos][character].Set(uint32(i))
					wordle.wordsContainingLetter[character].Set(uint32(i))
					found = true
				}
			}
			if !found {
				wordle.wordsNotContainingLetter[letter].Set(uint32(i))
			}
		}
	}

	return &wordle
}

func (w *Wordle) GetWord(i int) Word {
	return w.words[i]
}

func (w *Wordle) NWords() int {
	return len(w.words)
}

const (
	MISSING int = iota
	PRESENT
	CORRECT
)

func (w Word) Compare(soln Word) [5]int {
	var status [5]int

OUTER:
	for i, c := range w {
		for j, x := range soln {
			if c == x {
				if i == j {
					status[i] = CORRECT
					continue OUTER
				} else {
					status[i] = PRESENT
				}
			}
		}
	}

	return status
}

func (w *Wordle) Ambiguities(guesses []Word, soln Word) bitmap.Bitmap {
	var possible bitmap.Bitmap
	possible.Grow(uint32(len(w.words)))
	possible.Ones()

	for _, guess := range guesses {
		status := guess.Compare(soln)
		for letter, stat := range status {
			switch stat {
			case CORRECT:
				possible.And(w.wordsContainingLetterByPosition[letter][guess[letter]])
			case PRESENT:
				possible.And(w.wordsContainingLetter[guess[letter]])
			case MISSING:
				possible.And(w.wordsNotContainingLetter[guess[letter]])
			default:
			}
		}
	}
	return possible
}

type ComboProb struct {
	Combination []Word
	Probability float64
	Deduced     int
}

func (cp ComboProb) String() string {
	words := make([]string, len(cp.Combination))
	for i, w := range cp.Combination {
		words[i] = w.String()
	}
	joined := strings.Join(words, " + ")
	return fmt.Sprintf("%s = %f (%d deduced)", joined, cp.Probability, cp.Deduced)
}

func main() {

	flag.Parse()
	if flag.NArg() != 2 {
		log.Fatal("wordle requres two positional arguments: a file containing a list of possible solutions and a file containing a list of possible guesses")
	}

	solnFile, err := os.Open(flag.Arg(0))
	if err != nil {
		log.Fatal(err)
	}
	defer solnFile.Close()

	guessFile, err := os.Open(flag.Arg(1))
	if err != nil {
		log.Fatal(err)
	}
	defer guessFile.Close()

	start := []Word{}
	for _, word := range strings.Split(*startingWords, ",") {
		start = append(start, makeWordFromString(word))
	}

	solns := readWords(solnFile)
	guesses := readWords(guessFile)

	wordle := NewWordle(solns)

	combinations := wordCombinations(guesses, start, *nGuesses)                       // produce combinations
	unfiltered := calculateProbabilities(wordle, solns, *forceDisjoint, combinations) // multi-thread calculate solutions
	filtered := filterBest(unfiltered)                                                // merge

	for cp := range filtered {
		log.Print(cp)
	}
}

func calculateProbabilities(wordle *Wordle, solns []Word, disjoint bool, in <-chan []Word) <-chan ComboProb {
	out := make(chan ComboProb, 1000)
	filter := func(words []Word) bool {
		return true
	}
	if disjoint {
		filter = func(words []Word) bool {
			return disjointLetters(words)
		}
	}
	nsolns := float64(len(solns))
	go func() {
		var wg sync.WaitGroup
		for words := range in {
			wg.Add(1)
			go func(words []Word) {
				defer wg.Done()

				if !filter(words) {
					return
				}

				prob := 0.
				deduced := 0
				for _, solution := range solns {
					ambiguities := wordle.Ambiguities(words, solution).Count()
					if ambiguities == 1 {
						deduced++
					}
					prob += 1. / float64(ambiguities)
				}
				out <- ComboProb{Combination: words, Probability: prob / nsolns, Deduced: deduced}
			}(words)
		}
		wg.Wait()
		close(out)
	}()
	return out
}

func filterBest(in <-chan ComboProb) <-chan ComboProb {
	out := make(chan ComboProb, 1000)
	go func() {
		best := ComboProb{Combination: make([]Word, 0)}
		for cp := range in {
			if cp.Probability > best.Probability {
				if len(best.Combination) != len(cp.Combination) {
					best.Combination = make([]Word, len(cp.Combination))
				}
				copy(best.Combination, cp.Combination)
				best.Probability = cp.Probability
				best.Deduced = cp.Deduced
				out <- cp
			}
		}
		close(out)
	}()
	return out
}

func readWords(r io.Reader) []Word {
	words := make([]Word, 0)
	scanner := bufio.NewScanner(r)
	for scanner.Scan() {
		words = append(words, makeWord(scanner.Bytes()))
	}
	return words
}

func disjointLetters(ws []Word) bool {
	if len(ws) == 0 {
		return true
	}
	letters := make(map[byte]struct{})
	for i, w := range ws {
		for _, l := range w {
			letters[l] = struct{}{}
		}
		if len(letters) != (i+1)*5 {
			return false
		}
	}
	return true
}

func wordCombinations(ws []Word, start []Word, n int) <-chan []Word {
	if len(start)+n > 6 {
		panic("a maximum of only 6 guesses are allowed!")
	}
	out := make(chan []Word, 1000)
	go func() {
		numComb := combin.Binomial(len(ws), n)
		bar := pb.ProgressBarTemplate(pb.Full).Start(numComb)
		gen := combin.NewCombinationGenerator(len(ws), n)
		comb := make([]int, n)
		for gen.Next() {
			gen.Combination(comb)
			words := make([]Word, n+len(start))
			copy(words, start)
			for i, j := range comb {
				words[i+len(start)] = ws[j]
			}
			out <- words
			bar.Increment()
		}
		bar.Finish()
		close(out)
	}()
	return out
}
